concordium-consensus/src/Concordium/Scheduler.hs

```diff
@@ -83,48 +88,79 @@ import Prelude hiding (exp, mod)
-
-      return (iacc, cost)
+      
+      -- check if the transaction has been verified before and that the result is in the
+      -- transaction verification cache.
+      case HM.lookup hash (cache ^. normalTransactions) of
+        Just (NormalVerificationResult keysHash _) -> do
+          -- check that the keys for the account have not changed
+          currentKeys <- lift (TV.getAccountVerificationKeys $ snd iacc)
+          -- if the current account keys matches t
+          if getHash currentKeys == keysHash then return (iacc, cost)
+          -- keys changed the signature is no longer valid
+          else throwError $ Just IncorrectSignature
+        Nothing -> do
+          verRes <- lift (TV.verifyNormalTransaction meta)
+          if TV.isOk verRes then return (iacc, cost)
+          else throwError $ Just (mapTransactionVerificationResult verRes)
+
+-- |Maps transaction verification results into `FailureKind`s.
+-- Note. As the transaction verification results also contains
+-- successes for the three types of transactions; namely
+-- NormalTransaction's, ChainUpdate's and CredentialDeployments.
+-- These results are included to avoid a non-exausting pattern match in order to
+-- make it easier to work with later on if something should change.
+mapTransactionVerificationResult :: TV.VerificationResult -> FailureKind
+mapTransactionVerificationResult (TV.CredentialDeploymentDuplicateAccountRegistrationID regId) = DuplicateAccountRegistrationID regId
+mapTransactionVerificationResult TV.CredentialDeploymentInvalidIdentityProvider = NonExistentIdentityProvider undefined
+mapTransactionVerificationResult TV.CredentialDeploymentInvalidAnonymityRevokers = UnsupportedAnonymityRevokers
+mapTransactionVerificationResult TV.CredentialDeploymentInvalidSignatures = IncorrectSignature
+mapTransactionVerificationResult TV.CredentialDeploymentExpired = ExpiredTransaction
+mapTransactionVerificationResult TV.ChainUpdateEffectiveTimeBeforeTimeout = InvalidUpdateTime
+mapTransactionVerificationResult TV.ChainUpdateInvalidSignatures = IncorrectSignature
+mapTransactionVerificationResult TV.NormalTransactionDepositInsufficient = DepositInsufficient
+mapTransactionVerificationResult (TV.NormalTransactionInvalidSender aaddr) = UnknownAccount aaddr
+mapTransactionVerificationResult (TV.NormalTransactionInvalidNonce nonce) = NonSequentialNonce nonce
+mapTransactionVerificationResult TV.NormalTransactionInvalidSignatures = IncorrectSignature
+mapTransactionVerificationResult TV.ChainUpdateInvalidSequenceNumber = undefined -- todo
+-- todo: The successful ones are here to complete the pattern matching
+-- however they are unreachable as we execute these transactions.
+mapTransactionVerificationResult TV.CredentialDeploymentSuccess = undefined -- todo 
+mapTransactionVerificationResult (TV.ChainUpdateSuccess _ _) = undefined -- todo
+mapTransactionVerificationResult (TV.NormalTransactionSuccess _ _) = undefined -- todo
```

Comment:
to**do**

Reply:
Probably it would make sense to split the verificationResult into OK/NotOK groups and then you only need to map this for the NotOK ones.